hal/halx86/apic/apic.c

```diff
@@ -671,87 +858,470 @@ HalDisableSystemInterrupt(
 }
 
+VOID
+NTAPI
+HalpDispatchSoftwareInterrupt(_In_ KIRQL Irql,
+                              _In_ PKTRAP_FRAME TrapFrame)
+{
+    PHALP_PCR_HAL_RESERVED HalReserved;
+    ULONG EFlags;
+
+    EFlags = __readeflags();
+    KeGetPcr()->Irql = Irql;
+
+    if (!(EFlags & EFLAGS_INTERRUPT_MASK))
+        _enable();
+
+    HalReserved = (PHALP_PCR_HAL_RESERVED)KeGetPcr()->HalReserved;
+    if (Irql == APC_LEVEL)
+    {
+        HalReserved->ApcRequested = FALSE;
+        KiDeliverApc(0, 0, (PKTRAP_FRAME)TrapFrame);
+    }
+    else if (Irql == DISPATCH_LEVEL)
+    {
+        HalReserved->DpcRequested = FALSE;
+        KiDispatchInterrupt();
+    }
+    else
+    {
+        DbgBreakPoint();
+    }
+
+    if (!(EFlags & EFLAGS_INTERRUPT_MASK))
+        _disable();
+}
+
+VOID
+FASTCALL
+HalpCheckForSoftwareInterrupt(_In_ KIRQL NewIrql,
+                              _In_ PKTRAP_FRAME TrapFrame)
+{
+    PHALP_PCR_HAL_RESERVED HalReserved;
+    BOOLEAN ApcRequested;
+    BOOLEAN DpcRequested;
+
+    HalReserved = (PHALP_PCR_HAL_RESERVED)KeGetPcr()->HalReserved;
+
+    ApcRequested = HalReserved->ApcRequested;
+    DpcRequested = HalReserved->DpcRequested;
+
+    if (NewIrql)
+    {
+        if (NewIrql == APC_LEVEL && HalReserved->DpcRequested)
+        {
+            do
+            {
+                HalpDispatchSoftwareInterrupt(DISPATCH_LEVEL, TrapFrame);
+                KeSetCurrentIrql(APC_LEVEL);
+                HalReserved = (PHALP_PCR_HAL_RESERVED)KeGetPcr()->HalReserved;
+            }
+            while (HalReserved->DpcRequested);
+        }
+
+        return;
+    }
+
+    while (ApcRequested | DpcRequested)
+    {
+        if (DpcRequested)
+        {
+            HalpDispatchSoftwareInterrupt(DISPATCH_LEVEL, TrapFrame);
+        }
+        else
+        {
+            HalpDispatchSoftwareInterrupt(APC_LEVEL, TrapFrame);
+        }
+
+        KeSetCurrentIrql(PASSIVE_LEVEL);
+
+        HalReserved = (PHALP_PCR_HAL_RESERVED)KeGetPcr()->HalReserved;
+        ApcRequested = HalReserved->ApcRequested;
+        DpcRequested = HalReserved->DpcRequested;
+    }
+}
+
+BOOLEAN
+NTAPI
+HalBeginSystemInterrupt(_In_ KIRQL NewIrql,
+                        _In_ ULONG SystemVector,
+                        _Out_ PKIRQL OutOldIrql)
+{
+    PUCHAR pPrcbVector;
+    UCHAR Idx;
+    KIRQL OldIrql;
+
+    OldIrql = KeGetCurrentIrql();
+
+    if (OldIrql < HalpVectorToIRQL[(UCHAR)SystemVector >> 4])
+    {
+        *OutOldIrql = OldIrql;
+        KeSetCurrentIrql(NewIrql);
+
+        _enable();
+        return TRUE;
+    }
+
+    pPrcbVector = (PUCHAR)KeGetCurrentPrcb()->HalReserved;
+
+    Idx = pPrcbVector[0];
+    pPrcbVector[Idx + 1] = SystemVector;
+    pPrcbVector[0] = Idx + 1;
+
+    _enable();
+    return FALSE;
+}
+
+static
+VOID
+NTAPI
+HalpEndSystemInterrupt(_In_ PHAL_INTERRUPT_CONTEXT IntContext,
+                       _In_ KIRQL OldIrql)
+{
+    KIRQL Irql = HalpVectorToIRQL[IntContext->Vector >> 4];
+
+    if (Irql < KeGetCurrentIrql())
```

Comment:
This seem to be the only place where `Vector` field is used (for which you needed to change how the interrupts are processed in the kernel https://github.com/reactos/reactos/pull/3394/commits/420be339b28ea4ec178623593141bf366ac5ca12)

When does this happen? If an interrupt with a higher IRQL appeared while we're processing this one? What if the purpose of `HalpLowerIrqlHardwareInterrupts` then?

Reply:
In case the HAL must be NT compatible (no __REACTOS__) "Vector" is required. When the case is __REACTOS__ (like the current one), then theoretically it is possible. If you have an option for how to do this, then you can suggest it.

HalpLowerIrqlHardwareInterrupts(), as its name implies, is used to handle pending interrupts that were requested during the current interrupt (and whose IRQL was lower than the current interrupt IRQL).
If the incoming interrupt has a higher priority, then the processor is switched over to handle it.
If it is lower, then we need to remember it and process it later, after the completion of the current one.